lib/truncato/truncated_sax_document.rb

```diff
@@ -2,42 +2,92 @@
 require 'htmlentities'
 
 class TruncatedSaxDocument < Nokogiri::XML::SAX::Document
-  IGNORABLE_TAGS = %w(html head body)
+  IGNORABLE_TAGS = %w[html head body].freeze
 
-  SINGLE_TAGS = %w{br img}
+  # These don't have to be closed (which also impacts ongoing length calculations)
+  # http://www.456bereastreet.com/archive/201005/void_empty_elements_and_self-closing_start_tags_in_html/
+  VOID_TAGS = %w[area base br col command hr img input keygen link meta param source].freeze
 
-  attr_reader :truncated_string, :max_length, :max_length_reached, :tail,
-              :count_tags, :filtered_attributes, :filtered_tags, :ignored_levels
+  attr_reader :truncated_string,
+    :max_length,
+    :max_length_reached,
+    :tail,
+    :ignored_levels,
+    :something_has_been_truncated
 
   def initialize(options)
     @html_coder = HTMLEntities.new
-    capture_options options
-    init_parsing_state
+
+    @max_length = options[:max_length]
+    @tail = options[:tail]
+
+    @truncated_string = ""
+    @closing_tags = []
+    @estimated_length = 0
+    @max_length_reached = false
+    @ignored_levels = 0
   end
 
-  def start_element name, attributes
-    enter_ignored_level if filtered_tags.include?(name)
-    return if @max_length_reached || ignorable_tag?(name) || ignore_mode?
+  # This method is called when the parser encounters an open tag
+  def start_element(name, attributes)
+    return if @max_length_reached || ignorable_tag?(name)
+
+    string_to_add = opening_tag(name, attributes)
+
+    # Abort if there is not enough space to add the combined opening tag and (potentially) the closing tag
+    length_of_tags = overridden_tag_length(name, string_to_add)
+    if length_of_tags > remaining_length
+      enter_ignored_level
+      return
+    end
+
+    # Save the tag so we can push it on at the end
     @closing_tags.push name unless single_tag_element? name
```

Comment:
Seems like this makes an assumption that the `VOID_TAGS` would never be seen with a closing tag, but is that really true? Isn't it possible that some client out there somewhere closes one of those tags even though it does not need to? Is this enforced some other way? Or another way to ask the question is, what happens if it *does* happen to have a corresponding closing tag but we are not pushing it here onto the stack?